pyro/distributions/transforms/generalized_channel_permute.py

```diff
@@ -0,0 +1,164 @@
+    event_dim = 3
+
+    def __init__(self, channels=3):
+        super(GeneralizedChannelPermute, self).__init__(cache_size=1)
+        self.channels = channels
+
+        # Sample a random orthogonal matrix
+        W, _ = torch.qr(torch.randn(channels, channels))
+
+        # Construct the partially pivoted LU-form and the pivots
+        LU, pivots = W.lu()
+
+        # Convert the pivots into the permutation matrix
+        P, _, _ = torch.lu_unpack(LU, pivots)
+
+        # We register the permutation matrix so that the model can be serialized
+        self.register_buffer('permutation', P)
+
+        # NOTE: For this implementation I have chosen to store the parameters densely, rather than
+        # storing L, U, and s separately
+        self.LU = torch.nn.Parameter(LU)
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a
+        :class:`~pyro.distributions.TransformedDistribution` `x` is a sample from the base distribution (or the output
+        of a previous transform)
+        """
+
+        # Extract the lower and upper matrices from the packed LU matrix
+        U = self.LU.triu()
+        L = self.LU.tril()
+        L.diagonal(dim1=-2, dim2=-1).fill_(1)
+
+        # Perform the 2D convolution, using the weight
+        filters = (self.permutation @ L @ U)[..., None, None]
+        y = F.conv2d(x.view(-1, *x.shape[-3:]), filters)
+
+        return y.view_as(x)
+
+    def _inverse(self, y):
+        """
+        :param y: the output of the bijection
+        :type y: torch.Tensor
+
+        Inverts y => x.
+        """
+
+        """
+        NOTE: This method is equivalent to the following two lines. Using Tensor.inverse() would be
+        numerically unstable, however.
+
+        U = self.LU.triu()
+        L = self.LU.tril()
+        L.diagonal(dim1=-2, dim2=-1).fill_(1)
+        filters = (self.permutation @ L @ U).inverse()[..., None, None]
+        x = F.conv2d(y.view(-1, *y.shape[-3:]), filters)
+        return x.view_as(y)
+
+        """
+
+        # Do a matrix vector product over the channel dimension
+        # in order to apply inverse permutation matrix
+        y_flat = y.flatten(start_dim=-2)
+        LUx = (y_flat.unsqueeze(-3) * self.permutation.T.unsqueeze(-1)).sum(-2)
+
+        # Solve L(Ux) = P^1y
+        U = torch.triu(self.LU)
+        L = self.LU.tril()
+        L.diagonal(dim1=-2, dim2=-1).fill_(1)
+        Ux, _ = torch.triangular_solve(LUx, L, upper=False)
+
+        # Solve Ux = (PL)^-1y
+        x, _ = torch.triangular_solve(Ux, U)
+
+        # Unflatten x
+        return x.view_as(y)
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Calculates the elementwise determinant of the log Jacobian, i.e. log(abs(det(dy/dx))).
+        """
+
+        h, w = x.shape[-2:]
+        log_det = h * w * torch.log(torch.abs(torch.diag(self.LU))).sum()
```

Comment:
The `.diag()` and `.sum()` look suspicious. If you intend to support batching (e.g. for multi-particle SVI), you'll need to restrict the `.sum()` to relevant dimensions, e.g.
```py
log_det = h * w * self.LU.diagonal(dim1=-2, dim2=-1).abs().log().sum((-3, -2, -1))
```

Reply:
This will work for the non-conditional version of `GeneralizedChannelPermute`. I will revisit this point when implementing the conditional version!

Reply:
I.e. it will work correctly for batching of different sizes for the non-conditional transform. See the following line where the `log_det` factor multiplies a vector of ones in the batch shape.